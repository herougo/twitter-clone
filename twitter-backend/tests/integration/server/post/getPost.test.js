const createApp = require("../../../../src/server/app");
const DI_NAMES = require("../../../../src/server/dependency-injection/names");
const request = require('supertest');
const createMockLogger = require("../../../utils/mocks/mockLogger");
const mongoose = require("mongoose");
const { populateDatabase, clearDatabase } = require("../../../utils/database/changeContents");
const { DB_IDS } = require("../../../utils/database/ids");
const { sanitizePaths } = require("../../../utils/sanitization");

let app;
let diContainer;
const endpoint = '/post';

// run once before all suites in the file
beforeAll(async () => {
    const appData = await createApp({
        [DI_NAMES.logger]: createMockLogger()
    });
    app = appData.app;
    diContainer = appData.diContainer;
});

// run once after all suites in the file
afterAll(async () => {
    await mongoose.connection.close(); // neccessary to avoid a jest error
});

describe("GET /post/:postId endpoint", () => {
    // run before each "test"
    beforeEach(async () => {
        await clearDatabase(diContainer);
        await populateDatabase(diContainer);
    });

    const sendToEndpoint = async (param, loggedInUserId) => {
        let fullEndpoint = `${endpoint}/${param}`;
        if (loggedInUserId) {
            fullEndpoint += `?loggedInUserId=${loggedInUserId}`;
        }
        return await request(app).get(fullEndpoint).send();
    };

    const sanitizePost = (post) => {
        return sanitizePaths(
            post,
            ['createdDate', 'replyTo.createdDate'],
            ''
        );
    }

    const sanitizeBody = (body) => {
        return {
            post: sanitizePost(body.post),
            replies: body.replies.map(sanitizePost)
        }
    };

    test("Success (with reply)", async () => {
        const response = await sendToEndpoint(DB_IDS.mainPost, DB_IDS.mainUser);
        expect(response.statusCode).toBe(200);
        const sanitizedBody = sanitizeBody(response.body);
        expect(sanitizedBody).toMatchSnapshot();
    });

    test("Success (without reply)", async () => {
        const response = await sendToEndpoint(DB_IDS.mainReply, DB_IDS.mainUser);
        expect(response.statusCode).toBe(200);
        const sanitizedBody = sanitizeBody(response.body);
        expect(sanitizedBody).toMatchSnapshot();
    });

    test("Invalid PostId", async () => {
        const response = await sendToEndpoint(DB_IDS.missingPost, DB_IDS.mainUser);
        expect(response.statusCode).toBe(400);
        expect(response.body.errors.message).toEqual("GetPost: Invalid postId");
    });

    test("Missing loggedInUserId", async () => {
        const response = await sendToEndpoint(DB_IDS.mainPost);
        expect(response.statusCode).toBe(400);
        expect(response.body.errors.message).toEqual("GetPost: Missing loggedInUserId");
    });

    test("Missing postId", async () => {
        const response = await sendToEndpoint('');
        expect(response.statusCode).toBe(404);
    });
});