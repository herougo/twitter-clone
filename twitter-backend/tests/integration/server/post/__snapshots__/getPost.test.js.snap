// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /post/:postId endpoint Success (with reply) 1`] = `
{
  "post": {
    "author": {
      "name": "hi there",
      "username": "username",
    },
    "content": "I like dogs.",
    "createdDate": "",
    "id": "675351d698452e4fadc376c3",
    "numDislikes": 0,
    "numLikes": 0,
    "userInteraction": "",
  },
  "replies": [
    {
      "author": {
        "name": "hello you",
        "username": "follower",
      },
      "content": "Me too!",
      "createdDate": "",
      "id": "675351d698452e4fadc376c4",
      "numDislikes": 0,
      "numLikes": 0,
      "userInteraction": "",
    },
  ],
}
`;

exports[`GET /post/:postId endpoint Success (without reply) 1`] = `
{
  "post": {
    "author": {
      "name": "hello you",
      "username": "follower",
    },
    "content": "Me too!",
    "createdDate": "",
    "id": "675351d698452e4fadc376c4",
    "numDislikes": 0,
    "numLikes": 0,
    "replyTo": {
      "author": {
        "name": "hi there",
        "username": "username",
      },
      "content": "I like dogs.",
      "createdDate": "",
    },
    "userInteraction": "",
  },
  "replies": [],
}
`;
